WatchFaceKotlin/wear/src/main/java/com/example/android/wearable/watchfacekotlin/watchface/AnalogComplicationWatchFaceService.kt

```diff
@@ -0,0 +1,834 @@
+                    invalidate()
+                }
+            }
+        }
+
+        private fun registerReceiver() {
+            if (registeredTimeZoneReceiver) {
+                return
+            }
+            registeredTimeZoneReceiver = true
+            val filter = IntentFilter(Intent.ACTION_TIMEZONE_CHANGED)
+            registerReceiver(timeZoneReceiver, filter)
+        }
+
+        private fun unregisterReceiver() {
+            if (!registeredTimeZoneReceiver) {
+                return
+            }
+            registeredTimeZoneReceiver = false
+            unregisterReceiver(timeZoneReceiver)
+        }
+
+        /**
+         * Starts the coroutine/flow "timer" to animate the second hand every second when the
+         * watch face is active.
+         */
+        @InternalCoroutinesApi
+        private fun startSecondHandAnimation() {
+            if (!secondHandAnimationActive()) {
+                secondHandAnimationJob = watchfaceCoroutineScope.launch(Dispatchers.Main) {
+                    secondHandAnimationFlow().collect {
+                        // Redraws second hand one second apart (triggered by flow)
+                        invalidate()
+                    }
+                }
+            }
+        }
+
+        /**
+         * Stops the coroutine/flow "timer" when the watch face isn't visible (or is in ambient
+         * mode).
+         */
+        private fun stopSecondHandAnimation() {
+            if (secondHandAnimationJob?.isActive == true) {
+                secondHandAnimationJob?.cancel()
+            }
+        }
+
+        /**
+         * Second hand should only be visible in active mode.
+         */
+        private fun secondHandVisible(): Boolean {
+            return isVisible && !ambient
+        }
+
+        private fun secondHandAnimationActive(): Boolean {
+            return secondHandAnimationJob?.isActive ?: false
+        }
+
+        /**
+         * Creates a flow that triggers one second apart. It's used to animate the second hand when
+         * the watch face is in active mode.
+         */
+        private fun secondHandAnimationFlow(): Flow<Unit> = flow {
+            while (secondHandVisible()) {
+                val timeMs = System.currentTimeMillis()
+                Log.d(TAG, "timeMs: $timeMs")
+                val delayMs = INTERACTIVE_UPDATE_RATE_MS - (timeMs % INTERACTIVE_UPDATE_RATE_MS)
+                Log.d(TAG, "delayMs: $delayMs")
+                delay(delayMs)
+                // Emit next value
+                emit(Unit)
+            }
+        }
+    }
+
+    companion object {
+        private const val TAG = "AnalogWatchFace"
+
+        /*
+         * Update rate in milliseconds for interactive mode. We update once a second to advance the
+         * second hand.
+         */
+        private val INTERACTIVE_UPDATE_RATE_MS = TimeUnit.SECONDS.toMillis(1)
+
+        private const val HOUR_STROKE_WIDTH = 5f
```

Comment:
All of these are being used directly as pixels. These set be dip values, and converted to pixel based on the screen density.